geomesa-datastore-bundle/geomesa-datastore-processors/src/main/scala/org/geomesa/nifi/datastore/processor/records/GeoAvroRecordSetWriterFactory.scala

```diff
@@ -0,0 +1,111 @@
+  override def getSchema(map: util.Map[String, String], recordSchema: RecordSchema): RecordSchema = recordSchema
+
+  override def createWriter(componentLog: ComponentLog,
+                            recordSchema: RecordSchema,
+                            outputStream: OutputStream,
+                            map: util.Map[String, String]): GeoAvroRecordSetWriter = {
+    new GeoAvroRecordSetWriter(componentLog, recordSchema, outputStream, getConfigurationContext.getProperties)
+  }
+
+  override def getSupportedPropertyDescriptors: util.List[PropertyDescriptor] =
+    Collections.singletonList(GEOMETRY_COLUMNS)
+}
+
+object GeoAvroRecordSetWriterFactory {
+  val GEOMETRY_COLUMNS: PropertyDescriptor = new PropertyDescriptor.Builder()
+    .name("Geometry Columns")
+    .description("Comma-separated list of columns with geometries with type " +
+      "(and optionally the format, (Wkt, Wkb)).  " +
+      "Example: position:Point,line:LineString.  " +
+      "Note the first field is used as a the default geometry.")
+    .expressionLanguageSupported(ExpressionLanguageScope.VARIABLE_REGISTRY)
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+    .required(false).build
+
+  val TYPE_NAME: PropertyDescriptor = new PropertyDescriptor.Builder()
+    .name("SimpleFeature TypeName")
+    .description("The type name for the output.")
+    .expressionLanguageSupported(ExpressionLanguageScope.NONE)
+    .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+    .required(false).build
+}
+
+class GeoAvroRecordSetWriter(componentLog: ComponentLog, recordSchema: RecordSchema, outputStream: OutputStream, map: util.Map[PropertyDescriptor, String]) extends AbstractRecordSetWriter(outputStream) {
+  private val schemaAccessWriter = new SchemaNameAsAttribute()
+  // Wkt is assumed to be the default GeometryEncoding.
+  // TODO:  Extract this as a property?
+  private val encodings = getEncodings(map, GeometryEncoding.Wkt)
+  private val defaultGeometryColumn: Option[String] = Option(map.get(GEOMETRY_COLUMNS)).map(_.split(":")(0))
+  val typeName: Option[String] = Option(map.get(TYPE_NAME))
+
+  val geometryColumns: Seq[GeometryColumn] = encodings.map { case (k, v) =>
+    GeometryColumn(k, v.clazz, defaultGeometryColumn.isDefined && defaultGeometryColumn.get.equals(k))
+  }.toSeq
+
+  val recordConverterOptions: RecordConverterOptions = RecordConverterOptions(typeName, None, geometryColumns)
+  lazy val converter: SimpleFeatureRecordConverter =
+    SimpleFeatureRecordConverter(recordSchema, recordConverterOptions)
+
+  private val sft: SimpleFeatureType = converter.sft
+  val writer = new AvroDataFileWriter(outputStream, sft)
+
+  override def writeRecord(record: Record): util.Map[String, String] = {
+    val sf = record match {
+      case sfmr: SimpleFeatureMapRecord => sfmr.sf
+      case _: Record => converter.convert(record)
+    }
+    writer.append(sf)
+    schemaAccessWriter.getAttributes(recordSchema)
```

Comment:
is this always going to be the same, since it doesn't depend on the input? if so, could cache it. i admit i don't know what this method is actually supposed to return

Reply:
https://github.com/apache/nifi/blob/main/nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/AbstractRecordSetWriter.java#L58-L61
https://github.com/apache/nifi/blob/main/nifi-commons/nifi-record/src/main/java/org/apache/nifi/serialization/WriteResult.java#L42-L45

Evidently, it is a map of things to add to the FlowFile attributes...

I'm following examples like this: https://github.com/apache/nifi/blob/main/nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/WriteCSVResult.java#L150

I wonder why they do not memoize the result as you are suggesting.  Thoughts?

Reply:
yeah not sure... maybe just add some comments documenting this thread